File: src/main/java/org/usfirst/frc/team1458/lib/core/BaseRobot.java
package org.usfirst.frc.team1458.lib.core;

import edu.wpi.first.wpilibj.RobotState;
import edu.wpi.first.wpilibj.SampleRobot;
import org.usfirst.frc.team1458.lib.util.GlobalTeleUpdate;

import java.util.ArrayList;

/**
 * All robot classes should extend from this class.
 *
 * @author asinghani
 */
public abstract class BaseRobot extends SampleRobot implements AutoModeHolder {

	protected ArrayList<AutoMode> autoModes = new ArrayList<>();
	protected int selectedAutoMode = 0;

	/**
	 * Initialize the robot's hardware and basic configuration.
	 */
	public abstract void robotSetup();

	/**
	 * This method should be overridden and add autonomous modes (which extend AutoMode) using the method addAutoMode()
	 */
	public abstract void setupAutoModes();

	/**
	 * Initialize any robot configuration which is time-consuming and must be run on a separate thread.
	 */
	public abstract void threadedSetup();

	/**
	 * Setup for teleop (if necessary)
	 */
	public abstract void teleopInit();

	/**
	 * Called constantly during teleop period
	 */
	public abstract void teleopPeriodic();

	/**
	 * Run the robot's test mode.
	 */
	public abstract void runTest();

	/**
	 * Called when the robot is disabled.
	 */
	public abstract void robotDisabled();

	@Override
	public ArrayList<? extends AutoMode> getAutoModes() {
		return autoModes;
	}

	@Override
	public void setSelectedAutoModeIndex(int index) {
		this.selectedAutoMode = index;
	}

	@Override
	public int getSelectedAutoModeIndex() {
		return selectedAutoMode;
	}

	public final void addAutoMode(AutoMode autoMode) {
		autoModes.add(autoMode);
	}

	@Override
	protected final void robotInit() {
		robotSetup();
		setupAutoModes();
		if(autoModes.isEmpty()) {
			autoModes.add(new BlankAutoMode());
		}

	}

	@Override
	protected final void disabled() {
		robotDisabled();
	}

	@Override
	public final void autonomous() {
		AutoMode autoMode = autoModes.get(selectedAutoMode);
		if(autoMode != null) {
			autoMode.auto();
		}
	}

	@Override
	public final void operatorControl() {
		teleopInit();
		while(RobotState.isOperatorControl() && RobotState.isEnabled()) {
			teleopPeriodic();
			GlobalTeleUpdate.teleUpdate();
		}
	}

	@Override
	public final void test() {
		runTest();
	}
}

File: src/main/java/org/usfirst/frc/team1458/lib/core/BaseRobot.kt
package org.usfirst.frc.team1458.lib.core

import edu.wpi.first.wpilibj.SampleRobot
import org.usfirst.frc.team1458.lib.util.DataLogger
import org.usfirst.frc.team1458.lib.util.flow.WaitGroup
import org.usfirst.frc.team1458.lib.util.flow.delay
import org.usfirst.frc.team1458.lib.util.flow.go
import org.usfirst.frc.team1458.lib.util.flow.systemTimeMillis
import java.util.*

/**
 * All robot classes should extend from this class.
 *
 * @author asinghani
 */
abstract class BaseRobot : SampleRobot, AutoModeHolder {

    override val autoModes = ArrayList<AutoMode>()
        get

    override var selectedAutoModeIndex = 0

    constructor() : super()

    /**
     * Initialize the robot's hardware and basic configuration.
     */
    abstract fun robotSetup()

    /**
     * This method should be overridden and add autonomous modes (which extend AutoMode) using the method addAutoMode()
     */
    abstract fun setupAutoModes()

    /**
     * Initialize any robot configuration which is time-consuming and must be run on a separate thread.
     * This is not guaranteed to finish before the robot is enabled
     */
    abstract fun threadedSetup()

    /**
     * Setup for teleop (if necessary)
     */
    abstract fun teleopInit()

    /**
     * Called constantly during teleop period
     */
    abstract fun teleopPeriodic()

    /**
     * Run the robot's test mode.
     */
    abstract fun runTest()

    /**
     * Called when the robot is disabled.
     */
    abstract fun robotDisabled()

    fun addAutoMode(autoMode: AutoMode) {
        autoModes.add(autoMode)
    }

    override fun robotInit() {
        robotSetup()
        setupAutoModes()
        go { threadedSetup() }

        if (autoModes.isEmpty()) {
            autoModes.add(BlankAutoMode())
        }
    }

    override fun disabled() {
        robotDisabled()
    }

    override fun autonomous() {
        val autoMode = autoModes[selectedAutoModeIndex]
        autoMode?.auto()
    }

    override fun operatorControl() {
        teleopInit()
        while (super.isOperatorControl() && super.isEnabled()) {
            var lastStartMillis = systemTimeMillis
            DataLogger.currentIterationTimestamp = lastStartMillis
            teleopPeriodic()
            DataLogger.endTeleop()
            var lastEndMillis = systemTimeMillis

            var nextStartMillis : Double = lastStartMillis

            while(nextStartMillis < lastEndMillis || (nextStartMillis.toLong() % 20) != 0L) {
                nextStartMillis += 1
            }

            while(systemTimeMillis < nextStartMillis) {
                delay(1)
            }
        }
    }

    override fun test() {
        runTest()
    }
}